File: packages/sdk-node/test/enforcement-log.test.ts
import { describe, it, expect } from 'vitest';
import { RateGuard } from '../src/index.js';

const okResponse = () =>
  new Response(
    JSON.stringify({ model: 'gpt-4o', usage: { prompt_tokens: 400, completion_tokens: 100, total_tokens: 500 } }),
    { status: 200, headers: { 'content-type': 'application/json' } },
  );

function send(wrapped: typeof fetch, customer?: string) {
  return wrapped('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    body: JSON.stringify({ model: 'gpt-4o' }),
    ...(customer ? { headers: { 'X-RateGuard-Customer': customer } } : {}),
  });
}

describe('enforcement log', () => {
  it('records budget stops and freezes, newest first, with timestamps', async () => {
    const rg = new RateGuard({ preset: 'dev', tokenBudget: { hourLimit: 600 } });
    const wrapped = rg.wrapFetch({ estimatedTokens: 500, fetch: (async () => okResponse()) as typeof fetch });

    await send(wrapped, 'alice'); // 500 used
    await send(wrapped, 'alice'); // 1000 used
    await send(wrapped, 'alice'); // blocked -> token_budget_exceeded
    rg.freeze('bob');
    await send(wrapped, 'bob'); // frozen

    const events = rg.enforcementEvents(0);
    expect(events.length).toBeGreaterThanOrEqual(2);
    expect(events[0]!.type).toBe('frozen'); // newest first
    expect(events[0]!.customer).toBe('bob');
    expect(events.some((e) => e.type === 'token_budget_exceeded' && e.customer === 'alice')).toBe(true);
    expect(events[0]!.at).toMatch(/^\d{4}-\d{2}-\d{2}T/); // ISO timestamp
  });
});

File: packages/sdk-go/enforcement_log_test.go
package rateguard

import (
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestEnforcementLogRecordsBlocks(t *testing.T) {
	server := httptest.NewServer(openAIJSONHandler("gpt-4o", 400, 100))
	defer server.Close()
	sdk := New(Config{Preset: "dev", TokenBudgetPerHour: 600, EstimatedTokensPerRequest: 500})
	client := wrapForHost(t, sdk, server)

	send := func(customer string) {
		req, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions",
			strings.NewReader(`{"model":"gpt-4o"}`))
		if customer != "" {
			req.Header.Set("X-RateGuard-Customer", customer)
		}
		resp, err := client.Do(req)
		if err != nil {
			t.Fatalf("call failed: %v", err)
		}
		_, _ = io.ReadAll(resp.Body)
		_ = resp.Body.Close()
	}

	// alice burns her 600-token budget: the third call is blocked.
	send("alice")
	send("alice")
	send("alice")
	// bob is frozen by an operator.
	sdk.Freeze("bob")
	send("bob")

	events := sdk.EnforcementEvents(0)
	if len(events) < 2 {
		t.Fatalf("expected at least 2 enforcement events, got %d", len(events))
	}
	// Newest first: bob's freeze is the most recent block.
	if events[0].Type != "frozen" || events[0].Customer != "bob" {
		t.Errorf("newest event = %+v, want type=frozen customer=bob", events[0])
	}
	var foundBudget bool
	for _, e := range events {
		if e.Type == "token_budget_exceeded" && e.Customer == "alice" {
			foundBudget = true
		}
	}
	if !foundBudget {
		t.Error("expected a token_budget_exceeded event attributed to alice")
	}
}

func TestAdminEventsEndpoint(t *testing.T) {
	server := httptest.NewServer(openAIJSONHandler("gpt-4o", 10, 10))
	defer server.Close()
	sdk := New(Config{Preset: "dev"})
	sdk.Freeze("")
	client := wrapForHost(t, sdk, server)

	req, _ := http.NewRequest(http.MethodPost, "https://api.openai.com/v1/chat/completions",
		strings.NewReader(`{"model":"gpt-4o"}`))
	resp, _ := client.Do(req)
	_, _ = io.ReadAll(resp.Body)
	_ = resp.Body.Close()

	admin := sdk.AdminHandler()

	rr := httptest.NewRecorder()
	admin.ServeHTTP(rr, httptest.NewRequest(http.MethodGet, "/admin/events", nil))
	if rr.Code != http.StatusOK || !strings.Contains(rr.Body.String(), "frozen") {
		t.Errorf("GET /admin/events: code=%d body=%s", rr.Code, rr.Body.String())
	}

	rr = httptest.NewRecorder()
	admin.ServeHTTP(rr, httptest.NewRequest(http.MethodGet, "/admin/events?format=csv", nil))
	if ct := rr.Header().Get("Content-Type"); !strings.HasPrefix(ct, "text/csv") {
		t.Errorf("csv content-type = %q, want text/csv", ct)
	}
	if !strings.Contains(rr.Body.String(), "at,type,customer,provider,model,detail") {
		t.Errorf("csv missing header row: %s", rr.Body.String())
	}
}
